### ISSUE TITLE
Unable to run the publisher offline

### ISSUE THREAD

**Issue by cames-be** [HL7/fhir-ig-publisher#472]
TLDR; Disconnecting machine from internet and running the publisher offline results in the following error:

```
2022-07-19 13:45:31.495 [main] INFO  o.h.f.u.n.FilesystemPackageCacheManager [FilesystemPackageCacheManager.java:235] Failed to determine latest version of package hl7.terminology.r4 from server: http://packages.fhir.org
2022-07-19 13:45:33.286 [main] INFO  o.h.f.u.n.FilesystemPackageCacheManager [FilesystemPackageCacheManager.java:235] Failed to determine latest version of package hl7.terminology.r4 from server: https://packages2.fhir.org/packages
Error connecting to build server - running without build (build.fhir.org)
java.net.UnknownHostException: build.fhir.org
```

Our organization's policy is that our build machines can't connect out to the internet so we're trying to set up our build environment where all of our tooling has what it needs to run offline. The idea here is that we need to manually vet changes to dependencies before our builds can rely on them.  This helps to protect our org from bugs or attacks against our dependencies and also stabilizes our builds by reducing reliance on external systems.

I have taken the gzipped tarball files from my `%UserProfile%\.fhir\package-client` folder and moved them into a `packages-client` folder in my local project.  Then, assuming the jar file and everything else is cached locally, I issue the following command:

```
JAVA -jar "input-cache\publisher.jar" -ig . -packages "packages-client" -tx n/a
```

This builds the IG correctly.  If however, I disconnect my machine from the internet and attempt to build, I get the following errors:

```
Core Package hl7.fhir.r4.core#4.0.1
Terminology Cache is at ##path_to_project##\input-cache\txcache. 4 files in cache (00:22.421)
WARNING: Running without terminology server - terminology content will likely not publish correctly (00:22.427)
Load Package hl7.fhir.pubpack#0.1.1
Load Package hl7.fhir.xver-extensions#0.0.11
2022-07-19 13:45:31.495 [main] INFO  o.h.f.u.n.FilesystemPackageCacheManager [FilesystemPackageCacheManager.java:235] Failed to determine latest version of package hl7.terminology.r4 from server: http://packages.fhir.org
2022-07-19 13:45:33.286 [main] INFO  o.h.f.u.n.FilesystemPackageCacheManager [FilesystemPackageCacheManager.java:235] Failed to determine latest version of package hl7.terminology.r4 from server: https://packages2.fhir.org/packages
Error connecting to build server - running without build (build.fhir.org)
java.net.UnknownHostException: build.fhir.org
        at java.net.AbstractPlainSocketImpl.connect(Unknown Source)
        at java.net.PlainSocketImpl.connect(Unknown Source)
        at java.net.SocksSocketImpl.connect(Unknown Source)
        at java.net.Socket.connect(Unknown Source)
        at sun.security.ssl.SSLSocketImpl.connect(Unknown Source)
        at sun.net.NetworkClient.doConnect(Unknown Source)
        at sun.net.www.http.HttpClient.openServer(Unknown Source)
        at sun.net.www.http.HttpClient.openServer(Unknown Source)
        at sun.net.www.protocol.https.HttpsClient.<init>(Unknown Source)
        at sun.net.www.protocol.https.HttpsClient.New(Unknown Source)
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect(Unknown Source)
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
        at java.net.HttpURLConnection.getResponseCode(Unknown Source)
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(Unknown Source)
        at org.hl7.fhir.utilities.SimpleHTTPClient.get(SimpleHTTPClient.java:149)
        at org.hl7.fhir.utilities.SimpleHTTPClient.get(SimpleHTTPClient.java:121)
        at org.hl7.fhir.utilities.npm.FilesystemPackageCacheManager.loadFromBuildServer(FilesystemPackageCacheManager.java:672)
        at org.hl7.fhir.utilities.npm.FilesystemPackageCacheManager.checkBuildLoaded(FilesystemPackageCacheManager.java:661)
        at org.hl7.fhir.utilities.npm.FilesystemPackageCacheManager.getPackageUrlFromBuildList(FilesystemPackageCacheManager.java:566)
        at org.hl7.fhir.utilities.npm.FilesystemPackageCacheManager.fetchVersionTheOldWay(FilesystemPackageCacheManager.java:764)
        at org.hl7.fhir.utilities.npm.FilesystemPackageCacheManager.getLatestVersion(FilesystemPackageCacheManager.java:240)
        at org.hl7.fhir.igtools.publisher.Publisher.initializeFromIg(Publisher.java:2674)
        at org.hl7.fhir.igtools.publisher.Publisher.initialize(Publisher.java:2170)
        at org.hl7.fhir.igtools.publisher.Publisher.execute(Publisher.java:849)
        at org.hl7.fhir.igtools.publisher.Publisher.main(Publisher.java:10193)
Publishing Content Failed: Unable to resolve package id hl7.terminology.r4       (00:24.297)
                                                                                 (00:24.299)
Use -? to get command line help                                                  (00:24.300)
                                                                                 (00:24.302)
Stack Dump (for debugging):                                                      (00:24.304)
org.hl7.fhir.exceptions.FHIRException: Unable to resolve package id hl7.terminology.r4
        at org.hl7.fhir.utilities.npm.FilesystemPackageCacheManager.fetchVersionTheOldWay(FilesystemPackageCacheManager.java:770)
        at org.hl7.fhir.utilities.npm.FilesystemPackageCacheManager.getLatestVersion(FilesystemPackageCacheManager.java:240)
        at org.hl7.fhir.igtools.publisher.Publisher.initializeFromIg(Publisher.java:2674)
        at org.hl7.fhir.igtools.publisher.Publisher.initialize(Publisher.java:2170)
        at org.hl7.fhir.igtools.publisher.Publisher.execute(Publisher.java:849)
        at org.hl7.fhir.igtools.publisher.Publisher.main(Publisher.java:10193)
```

We have a need to be able to give a known set of packages (including the core package) and have the publisher operate off of them without needing to access the internet.

**Comment by grahamegrieve** [comment]
That error is the first of many issues that you will run into. So firstly, there's a hard dependency on tx.fhir.org that you won't be able to remove, but you can run your own local copy of that in a way that's entirely disconnected from the internet. But as you have found, there is ongoing dependency on the package servers as well. And package-client is no longer used, so that won't help

We are considering providing a package server subsystem that's entirely file based, and that would help. Somewhat.

But in the end, this is a design question, and it will cost us to set up and maintain a test infrastructure that runs entirely off line, which we would need in order to be willing to call this supported functionality. (mainly it will cost us in time, but all my experience tells me it won't be trivial). And this will set behind other higher priorities on our task list

**Comment by cames-be** [comment]
That depends on how you are attempting to test. If you're just trying to test locally you can disable the network. I was able to reproduce the issue just unplugging my ethernet cable and disabling wifi. 

If you're looking for a unit test, the best way is to create an interface that can be mock'd. This would also allow for easier future enhancements such as using a self-hosted terminology server instead of the default `tx.fhir.org` that is currently required. 

The other option that sounds more like what you're looking for is an integration test. The best way would be to isolate inside a container such as Docker where the network access could be disabled using a command such as `docker run --network none`. I can provide a Dockerfile for this if needed.

**Comment by joethe** [comment]
I'm running into the same problem. I read the publisher documentation here: 
> -packages: (optional) directory containing packages (tarball/gzips with implementation guide-associated artifacts with extension .tgz) to be used for packages to load without using package management infrastructure

(https://confluence.hl7.org/display/FHIR/IG+Publisher+Documentation)

and assumed that I would be able to build a docker image containing _everything_ necessary to build an IG without any network access.

On a related note, is there documentation somewhere listing all of the online resources that the publisher requires / attempts to use?

**Comment by lmckenzi** [comment]
For IG publication to be accurate, it will need access to a terminology server.  It's possible to run without one, but rendering of terminology artifacts and elements is likely to be incomplete.  So it's intended as a temporary measure, not suitable to final publications.

**Comment by joethe** [comment]
> It's possible to run without one, but rendering of terminology artifacts and elements is likely to be incomplete.

Do you mean that the IG may be based on outdated information if run without a terminology server, or that certain parts of the IG simply will not be built at all without access to a terminology server (there is no way to provide that information to the publisher offline)?
Is it possible to pin an IG to specific versions of all of its dependencies to get a reproducible IG build?

**Comment by grahamegrieve** [comment]
it's possible, sometimes, but way more complicated that you hope it to be, and a recipe for getting it wrong. None of those testing approaches work for me so we'll just have to accept some risk around network access. There is a parameter to stop network access but that doesn't mean that things will work properly when it's used. It's effectively experimental for now

**Comment by lmckenzi** [comment]
No further discussion in the last year, so presuming we can call this 'closed'.  (I.e. offline publishing is now working as well as we hope for it to.)
